### PR TITLE
Replication documentation

### PR DESCRIPTION
**Pull request by aenario**
This is a big big wip.

Can we figure out an unified sharing/replication/permissions model that we can explain in 5min or less ?
@nono @jinroh @Gara64 @Ljinod 

:heart:

### CODE REVIEW

File: docs/replication.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+This could be done with a rsync-like mechanism: Rsync essentially works by computing a signature of the files on target, sending this signature to source. On source, the file signature is used to compute a diff and then send only the changed chunks of the file to target. The chunks are stored on a tmp file on disk, so subsequent rsync does not need to resend the whole file.
+
+The difficulty will be in maintaining synchronization of the couchdb & vfs state while synchronizing between hosts with two different replication algorithm. The MD5 stored in couchdb should help to ensure correctness. The already-planed *update-couchdb-to-vfs-state* function too.
+
+We might need a zsync route too.
+
+Some useful packages:
+- Pure go rsync implementation : https://godoc.org/github.com/smtc/rsync
+- Go bindings for librsync : https://github.com/silvasur/golibrsync
+- A package to sync folder in FS or Swift: http://rclone.org/ (not rsync based, always whole file transfer)
```
Comment (nono) [inline docs/replication.md]: I'd be really happy to make a pull request to rclone to support cozy cloud VFS

File: docs/replication.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+3. The replicator `POST :target/_revs_diff` to get a list of all revisions of the changed documents the target does not know.
+4. The replicator `GET :source/:docid?open_revs=['rev1', ...]` several times to retrieve the missing documents revisions.
+5. The replicator `POST :target/_bulk_docs` with these documents revisions.
+6. Store the new last sequence number
+
+Repeat 2-5 until there is no more changes.
+
+## Details
+
+- In step 5, the replicator can also attempt to `PUT :target/:docid` if doc are too heavy, but this should not happens in cozy-stack considering there wont be attachment in couchdb.
```
Comment (nono) [inline docs/replication.md]: Maybe for emails ;-)

File: docs/replication.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+To be a target of replication, the stack need to support the following routes:
+
+- `POST :target/_revs_diff` takes a list of ID -> Rev and returns which one are missing.
+- `POST :target/_bulk_docs` create several documents at once
+- `POST :target/_ensure_full_commit` ensure the documents are written to disk. This is useless if couchdb is configured without delayed write (default), but remote couchdb will call, so the stack should return the expected 201
+
+In both case, we need to support
+
+- `PUT    :both/_local/:revdocid` to store the current sequence number.
+
```
Comment (nono) [inline docs/replication.md]: I think it should be nice to also support [`_all_docs`](http://docs.couchdb.org/en/2.0.0/api/database/bulk-api.html#db-bulk-docs). It is currently used by cozy-mobile to speed up the synchronization. I think pouchdb try to use it for replication, and has some fallbacks when it's not available.

File: docs/replication.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+- Have `_ensure_full_commit` [(Doc)](http://docs.couchdb.org/en/2.0.0/api/database/compact.html#db-ensure-full-commit) returns 201
+
+This will cover documents part of the Devices use case.
+
+With a way to list doctype, this could also cover Backup and Moving.
+
+
+## Continuous replication
+
+It is impossible to implement it by simply proxying to couchdb (see unlimited inactive users). But neither the Device nor the Backup option need open continuous replication.
```
Comment (nono) [inline docs/replication.md]: Cozy-desktop uses continuous replication (well, only a continuous changes-feed), so I disagree here to say that Device doesn't need open continuous replication.

File: docs/replication.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+## Realtime
+
+One of the cool feature of cozy apps was how changes are sent to the client in realtime through websocket. However this have a cost : every cozy and every apps open in a browser, even while not used keep one socket open on the server.
+
+Several technology can be used for this : Websocket, SSE or COMET like long-polling. Goroutines makes all solution similar performances wise. COMET is a hack, websocket seems more popular and tested (x/net/websocket vs html5-sse-example). SSE has some limitations (can't set header...)
+
+Depending on benchmarking, we can do some optimization on the feed :
+- close feeds when the user is not on screen
+- multiplex different applications' feed, so each open cozy will only use one socket to the server. And then use a [SharedWorker](https://developer.mozilla.org/en/docs/Web/API/SharedWorker) or postMessage to pass the events to each application javascript context (either iframe or browser tabs).
```
Comment (nono) [inline docs/replication.md]: The apps won't be on the same domains and we will make them not embeddable in iframes for security reasons. I don't think this let us some options to mutualize connections between apps. But we can see that later, it's only optimizations.

File: docs/replication.md (hunk 2 above)
Comment (aenario) [inline docs/replication.md]: On the things I'd do differently with emails if I were to start a new, is to consider the content as a text/plain or text/html attachment. It is really inefficient to pass the  whole content just to see the messages list.

File: docs/replication.md (hunk 3 above)
Comment (aenario) [inline docs/replication.md]: Added in 23a8580

**Comment by nono** [comment]
LGTM. @jinroh are you okay too to merge this PR?

**Pull request merged**